docs/proposals/improve-status-in-CAPI-resources.md

```diff
@@ -0,0 +1,921 @@
+| `Phase` (deprecated)          | `BackCompatibilty.Phase` (renamed) (deprecated)          | (removed)                                  |
+| `LastUpdated` (deprecated)    | `BackCompatibilty.LastUpdated` (renamed) (deprecated)    | (removed)                                  |
+| `FailureReason` (deprecated)  | `BackCompatibilty.FailureReason` (renamed) (deprecated)  | (removed)                                  |
+| `FailureMessage` (deprecated) | `BackCompatibilty.FailureMessage` (renamed) (deprecated) | (removed)                                  |
+| `Conditions`                  | `BackCompatibilty.Conditions` (renamed) (deprecated)     | (removed)                                  |
+| `ExprimentalConditions` (new) | `Conditions` (renamed)                                   | `Conditions`                               |
+| other fields...               | other fields...                                          | other fields...                            |
+
+Notes:
+- The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).
+  Fields in this struct are used for supporting down conversions, thus providing users relying on v1beta1 APIs additional buffer time to pick up the new changes. 
+
+##### Machine (New)Conditions
+
+| Condition              | Note                                                                                                                                                                                                                                                                        |
+|------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `Available`            | True if at the machine is Ready for at least MinReady seconds, as defined by the Machine's owner resource                                                                                                                                                                   |
+| `Ready`                | True if Machine's `BootstrapSecretReady`, `InfrastructureReady`, `NodeHealthy` and `HealthCheckSucceeded` (if present) are true; if other conditions are defined in `spec.readinessGates` are defined, those conditions should be true as well for the Machine to be ready. |
+| `UpToDate`             | True if the Machine spec matches the spec of the Machine's owner resource, e.g KubeadmControlPlane or MachineDeployment                                                                                                                                                     |
```

Comment:
In which cases the Machine spec wouldn't match and who's gonna do the checking?

Reply:
The idea is that it will be up to the Machine's owner controllers to set this condition.

E.g. With a machine deployment, after changing spec, all the existing machines - under a now old MachineSet - will get UpToDate=False, while the replacement machines created under the new current MachineSet will be created with UpToDate=True.

Also, this is something that opens up to track in place updates we are discussing in the working group

Reply:
\* I think the source of truth for the UpToDate condition has to come from the owning controller (whoever that might be). Not sure if that controller should also directly write the condition.

For example we could also just let the owning controller mark the resource as outdated and then the Machine controller writes the condition. We already ~ use this pattern with the revision annotations. Just might be better to have a clear standardized way to signal that a Machine is outdated. This would also provide a clear API surface for folks to signal a Machine is outdated, e.g. for other control plane providers. We shouldn't have some sort of "internal" API with KCP.

Reply:
@fabriziopandini IIRC the idea going forward was also that UpToDate is false if an in-place upgrade was pending. I think it would be really good if the Machine controller is writing this condition. It feels like a mix of responsibilities if the owning controller has to figure out if an in-place upgrade is (still) ongoing for a Machine

Reply:
Resolving this thread, hopefully this is clear now